### PR TITLE
windows: add invert ANSI (7 / 27) emulation

### PR DESCRIPTION
**Pull request by Qix-**
Since `COMMON_LVB_REVERSE_VIDEO` is not supported on any modern windows system, it is used as a flag and emulated through the bitwise flipping of fg/bg to create the inverse (reverse video) graphics in the console.

Uses the standard ANSI codes `[7;` and `[27;` for negative/positive.

### CODE REVIEW

**Comment by Qix-** [comment]
Done.

File: src/win/tty.c [hunk 1]
```diff
@@ -1238,6 +1258,23 @@ static int uv_tty_set_style(uv_tty_t* handle, DWORD* error) {
...
 
+  if (inverse != 0) {
+    if (inverse > 0) {
+      info.wAttributes |= COMMON_LVB_REVERSE_VIDEO;
+    } else {
+      info.wAttributes &= ~COMMON_LVB_REVERSE_VIDEO;
+    }
+  }
+
+  if ((info.wAttributes & COMMON_LVB_REVERSE_VIDEO) > 0) {
```
Comment (indutny) [inline src/win/tty.c:1272]: Is there any way to deduplicate it by macro or something like that?

File: src/win/tty.c [hunk 2]
```diff
@@ -1158,6 +1161,10 @@ static int uv_tty_set_style(uv_tty_t* handle, DWORD* error) {
       /* Background bright on */
       bg_bright = 1;
 
+    } else if (arg == 7) {
+      /* Inverse: on (negative) */
+      inverse = 1;
```
Comment (saghul) [inline src/win/tty.c:1174]: This reads a bit odd. "on (negative)" yet the value is positive. Maybe you can reword the comment a bit?

File: src/win/tty.c [hunk 3]
```diff
@@ -1166,6 +1173,10 @@ static int uv_tty_set_style(uv_tty_t* handle, DWORD* error) {
       /* Background bright off */
       bg_bright = 0;
 
+    } else if (arg == 27) {
+      /* Inverse: off (positive) */
+      inverse = -1;
```
Comment (saghul) [inline src/win/tty.c]: Ditto.

File: src/win/tty.c [hunk 4]
```diff
@@ -1126,6 +1126,7 @@ static int uv_tty_set_style(uv_tty_t* handle, DWORD* error) {
 
   char fg_color = -1, bg_color = -1;
   char fg_bright = -1, bg_bright = -1;
+  char inverse = 0;
```
Comment (piscisaureus) [inline src/win/tty.c]: The pattern used in this function is that if `fg_color`, `bg_bright` etc remain unchanged their value is -1.
You should do the same for `inverse`: initialize to -1 and then change the value to either 0 or 1 if it has to change.

File: src/win/tty.c (hunk 3 above)
Comment (piscisaureus) [inline src/win/tty.c]: So this should become `inverse = 0`.

File: src/win/tty.c [hunk 5]
```diff
@@ -1198,7 +1209,7 @@ static int uv_tty_set_style(uv_tty_t* handle, DWORD* error) {
   }
 
   if (fg_color == -1 && bg_color == -1 && fg_bright == -1 &&
-      bg_bright == -1) {
+      bg_bright == -1 && inverse == 0) {
```
Comment (piscisaureus) [inline src/win/tty.c]: inverse == -1

File: src/win/tty.c (hunk 1 above)
Comment (Qix-) [inline src/win/tty.c:1272]: Sure, what is the preference on location of macro defines relative to the function?

File: src/win/tty.c (hunk 2 above)
Comment (Qix-) [inline src/win/tty.c:1174]: Yea I thought I'd get such a response. I will update it to `-1`=unchanged, `0`=normal, `1`=inverted.

File: src/win/tty.c (hunk 3 above)
Comment (Qix-) [inline src/win/tty.c]: Yep; see [my comment above](https://github.com/joyent/libuv/pull/1243#r11830283).

**Comment by Qix-** [comment]
Updated the values to more closely follow the convention of the other flag values.

Just waiting on a response from @indutny about a possible macro and its location.

File: src/win/tty.c (hunk 1 above)
Comment (indutny) [inline src/win/tty.c:1272]: Right before the function should be fine ;)

**Comment by Qix-** [comment]
Not entirely sure why GitHub isn't showing the diff'd lines, but...
- [1120-1125](https://github.com/Qix-/libuv/blob/8aefacc68b255526566b156ff658f3622b8de799/src/win/tty.c#L1120-L1125): macro definition
- [1229](https://github.com/Qix-/libuv/blob/8aefacc68b255526566b156ff658f3622b8de799/src/win/tty.c#L1229): initial flip
- [1271](https://github.com/Qix-/libuv/blob/8aefacc68b255526566b156ff658f3622b8de799/src/win/tty.c#L1271): secondary flip

@indutny, I went ahead and put the macro right before the function. I assume that is where you meant.

EDIT: Is the diff issue (plus the fact Github thinks I've changed thousands of lines) a line ending issue?

**Comment by saghul** [comment]
All the line endings have been changed so the diff looks like it replaces the whole file. Please stick to Unix style line endings.

**Comment by Qix-** [comment]
No problem, text editor settings must have been changed. I'll fix it.

**Comment by Qix-** [comment]
@saghul Fixed line endings. :)

File: src/win/tty.c [hunk 6]
```diff
@@ -1117,6 +1117,12 @@ static int uv_tty_clear(uv_tty_t* handle, int dir, char entire_screen,
   return 0;
 }
 
+#define FLIP_FGBG \
```
Comment (saghul) [inline src/win/tty.c]: maybe use the do..while(0) construct here

Comment (Qix-) [inline src/win/tty.c]: Done :)

File: src/win/tty.c [hunk 7]
```diff
@@ -1117,6 +1117,14 @@ static int uv_tty_clear(uv_tty_t* handle, int dir, char entire_screen,
...
+    do { \
```
Comment (indutny) [inline src/win/tty.c]: Pleas align `\` at 79 column

**Comment by indutny** [comment]
LGTM after fixing few nits

**Comment by Qix-** [comment]
@indutny Aligned :) Thank you for the review!

**Comment by saghul** [comment]
@Qix- LGTM too. Can you squash and rebase?

**Comment by Qix-** [comment]
@saghul Done.

**Comment by saghul** [comment]
Landed in https://github.com/joyent/libuv/commit/e56717ae988416ddaebadbb1c23c07f466bab046. Thanks @Qix-!

**Pull request closed without merging**